airflow/providers/ftp/operators/ftp.py

```diff
@@ -0,0 +1,135 @@
+                create_intermediate_dirs=True,
+                dag=dag
+            )
+
+    """
+
+    template_fields: Sequence[str] = ('local_filepath', 'remote_filepath')
+
+    def __init__(
+        self,
+        *,
+        ftp_conn_id: str,
+        local_filepath: str | list[str],
+        remote_filepath: str | list[str],
+        operation: str = FTPOperation.PUT,
+        create_intermediate_dirs: bool = False,
+        **kwargs,
+    ) -> None:
+        super().__init__(**kwargs)
+        self.ftp_conn_id = ftp_conn_id
+        self.operation = operation
+        self.create_intermediate_dirs = create_intermediate_dirs
+
+        self.local_filepath_was_str = False
+        if isinstance(local_filepath, str):
+            self.local_filepath = [local_filepath]
+        else:
+            self.local_filepath = local_filepath
+
+        if isinstance(remote_filepath, str):
+            self.remote_filepath = [remote_filepath]
+        else:
+            self.remote_filepath = remote_filepath
+
+        if len(self.local_filepath) != len(self.remote_filepath):
+            raise ValueError(
+                f'{len(self.local_filepath)} paths in local_filepath '
+                f'!= {len(self.remote_filepath)} paths in remote_filepath'
+            )
+
+        if self.operation.lower() not in [FTPOperation.GET, FTPOperation.PUT]:
+            raise TypeError(
+                f"Unsupported operation value {self.operation}, "
+                f"expected {FTPOperation.GET} or {FTPOperation.PUT}."
+            )
+
+    @cached_property
+    def hook(self) -> FTPHook:
+        """Create and return an FTPHook."""
+        return FTPHook(ftp_conn_id=self.ftp_conn_id)
+
+    def execute(self, context: Any) -> str | list[str] | None:
+        file_msg = None
+        try:
+            for local_filepath, remote_filepath in zip(self.local_filepath, self.remote_filepath):
+                if self.operation.lower() == FTPOperation.GET:
+                    local_folder = os.path.dirname(local_filepath)
+                    if self.create_intermediate_dirs:
+                        Path(local_folder).mkdir(parents=True, exist_ok=True)
+                    file_msg = f"from {remote_filepath} to {local_filepath}"
+                    self.log.info("Starting to transfer %s", file_msg)
+                    self.hook.retrieve_file(remote_filepath, local_filepath)
+                else:
+                    remote_folder = os.path.dirname(remote_filepath)
+                    if self.create_intermediate_dirs:
+                        self.hook.create_directory(remote_folder)
+                    file_msg = f"from {local_filepath} to {remote_filepath}"
+                    self.log.info("Starting to transfer file %s", file_msg)
+                    self.hook.store_file(remote_filepath, local_filepath)
+        except Exception as e:
+            raise AirflowException(f"Error while transferring {file_msg}, error: {str(e)}")
```

Comment:
Minor nit:
Catching general `Exception` is a slight code-smell (most IDEs will yell at you for doing it, for exampole) and I'm not sure throwing a new `AirflowException` adds much here, you'll get the `During handling of the above exception, another exception occurred:` double exception output which I find just confuses a lot of people if the second exception isn't actually a problem.

If you just want to add more context you can log an error message in a `finally` block and then just let the original exception bubble up.

P.S. You could also use the py3 syntax `raise <new exception> from <old exception>` but you still get the double stacked exception output in that case, but it's now logged as `The above exception was the direct cause of the following exception:` which I still think causes more confusion than it solves.

Reply:
That is a good point. I think I originally just wanted to add context for the reader by telling them that the error occurred during the `"transfer of {file_msg}"`. But I think this context is unnecessary as the function already logs the `file_msg` before the calls to `self.hook.store_file` or `self.hook.retrieve_file`. So now I just remove the `try-except`, as it is unnecessary.